src/transfers-history/adapters/web3/SpokePoolEventsQuerier.ts

```diff
@@ -0,0 +1,111 @@
+import { SpokePool } from "@across-protocol/contracts-v2";
+import { TypedEvent, TypedEventFilter } from "@across-protocol/contracts-v2/dist/typechain/common";
+import { getSamplesBetween } from "../../../utils";
+import { Web3Error, Web3ErrorCode } from "./model";
+import { Logger } from "../logger";
+
+const DEFAULT_BLOCK_RANGE = 100_000;
+
+/**
+ * Interface implemented by classes that fetch contract events from the SpokePool contracts
+ */
+export interface ISpokePoolContractEventsQuerier {
+  getFundsDepositEvents: (from: number, to: number, depositorAddr?: string) => Promise<TypedEvent<any>[]>;
+  getFilledRelayEvents: (from: number, to: number, depositorAddr?: string) => Promise<TypedEvent<any>[]>;
+}
+
+/**
+ * Class that wraps the queryFilter calls and has the ability to resize the block range
+ * in order to comply with the eventual node restrictions in terms of the block range or the
+ * length of the response
+ */
+export class SpokePoolEventsQuerier implements ISpokePoolContractEventsQuerier {
+  constructor(private spokePool: SpokePool, private blockRangeSize?: number, private logger?: Logger) {}
+
+  public async getFundsDepositEvents(from: number, to: number, depositorAddr?: string): Promise<TypedEvent<any>[]> {
+    return this.getEvents(from, to, this.getDepositEventsFilters(depositorAddr));
+  }
+
+  public async getFilledRelayEvents(from: number, to: number, depositorAddr?: string): Promise<TypedEvent<any>[]> {
+    return this.getEvents(from, to, this.getFilledRelayEventsFilter(depositorAddr));
+  }
+
+  private getFilledRelayEventsFilter(depositorAddr?: string) {
+    if (depositorAddr) {
+      return this.spokePool.filters.FilledRelay(
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        depositorAddr,
+        undefined,
+        undefined
+      );
+    }
+    return this.spokePool.filters.FilledRelay();
+  }
+
+  private getDepositEventsFilters(depositorAddr?: string) {
+    if (depositorAddr) {
+      return this.spokePool.filters.FundsDeposited(
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        undefined,
+        depositorAddr.toLowerCase()
+      );
+    }
+    return this.spokePool.filters.FundsDeposited();
+  }
+
+  private async getEvents(
+    from: number,
+    to: number,
+    filters: TypedEventFilter<TypedEvent<any>[], any>
+  ): Promise<TypedEvent<any>[]> {
```

Comment:
Can we make this function generic so it will pass through these types? Or are the typechain types not specific enough to do that?

Reply:
I tried something like this, but it errors 

```ts
export class SpokePoolEventsQuerier<T extends TypedEvent<any>>  {
    async getEvents() {
        let events: T[] = [];
        const newEvents = await this.spokePool.queryFilter(filters, blockStart, blockEnd);
        events.push(...newEvents);
        // ^^ see screenshot for error
    }
}
```
<img width="1286" alt="Screenshot 2022-04-01 at 02 56 49" src="https://user-images.githubusercontent.com/89395931/161168900-4595910a-15a2-4182-bbeb-3ef86082a097.png">